File: 알고리즘 분류/다이나믹 프로그래밍/14501. 퇴사/퇴사.py
import sys

input = sys.stdin.readline

n = int(input())
data = [list(map(int, input().split())) for _ in range(n)]
dp = [0] * (n + 1)

for i in range(n - 1, -1, -1):
  if data[i][0] + i > n:
    dp[i] = dp[i + 1]
  else:
    dp[i] = max(dp[i + 1], data[i][1] + dp[data[i][0] + i])

print(dp[0])

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;

public class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine());
        int[][] data = new int[n][2];
        
        for (int i = 0; i < n; i++) {
            String[] input = br.readLine().split(" ");
            
            data[i][0] = Integer.parseInt(input[0]);
            data[i][1] = Integer.parseInt(input[1]);
        }
        
        int[] dp = new int[n + 1];
        for (int i = n - 1; i >= 0; i--) {
            if (data[i][0] + i > n) {
                dp[i] = dp[i + 1];
            } else {
                dp[i] = Math.max(dp[i + 1], dp[i] + data[i][1] + dp[data[i][0] + i]);
            }
        }
        System.out.println(dp[0]);
        
        br.close();
    }
}